db: inconsistent batch count panic

Just encountered another occurrence of the `inconsistent batch count` panic:

```
panic: pebble: inconsistent batch count: 4401699 vs 4401702
```

Initial analysis seems to show a truncated log record:

```
~ cockroach debug pebble wal dump cockroach/000538.log
...
47742712(253) seq=4401699 count=3
EOF
```

This looks reminiscent of #540. Was the fix in #543 insufficient? Or is something else going on?

--- a/internal/record/record.go
+++ b/internal/record/record.go
@@ -273,6 +273,9 @@ func (r *Reader) nextChunk(wantFirst bool) error {
 		}
 		n, err := io.ReadFull(r.r, r.buf[:])
 		if err != nil && err != io.ErrUnexpectedEOF {
+			if err == io.EOF && !wantFirst {
+				return io.ErrUnexpectedEOF
+			}
 			return err
 		}
 		r.begin, r.end, r.n = 0, 0, n

--- a/internal/record/record_test.go
+++ b/internal/record/record_test.go
@@ -732,7 +732,7 @@ func TestSeekRecord(t *testing.T) {
 	if err == nil {
 		t.Fatalf("Seek past the end of a file didn't cause an error")
 	}
-	if err != io.EOF {
+	if err != io.ErrUnexpectedEOF {
 		t.Fatalf("Seeking past EOF raised unexpected error: %v", err)
 	}
 	r.recover() // Verify recovery works.
@@ -930,6 +930,21 @@ func TestRecycleLog(t *testing.T) {
 			t.Fatalf("%d: expected EOF, but found %v", i, err)
 		}
 	}
+}
+
+func TestTruncatedLog(t *testing.T) {
+	backing := make([]byte, 2*blockSize)
+	w := NewLogWriter(bytes.NewBuffer(backing[:0]), base.FileNum(1))
+	// Write a record that spans 2 blocks.
+	_, err := w.WriteRecord(bytes.Repeat([]byte("s"), blockSize+100))
+	require.NoError(t, err)
+	require.NoError(t, w.Close())
+	// Create a reader only for the first block.
+	r := NewReader(bytes.NewReader(backing[:blockSize]), base.FileNum(1))
+	rr, err := r.Next()
+	require.NoError(t, err)
+	_, err = ioutil.ReadAll(rr)
+	require.EqualValues(t, err, io.ErrUnexpectedEOF)
 }
 
 func TestRecycleLogWithPartialBlock(t *testing.T) {

